lib/logthrdestdrv.c

```diff
@@ -201,6 +201,73 @@ _rewind_batch(LogThreadedDestDriver *self)
   self->batch_size = 0;
 }
 
+static void
+_process_result(LogThreadedDestDriver *self, gint result, LogMessage *triggering_message)
+{
+  switch (result)
+    {
+    case WORKER_INSERT_RESULT_DROP:
+      msg_error("Message(s) dropped while sending message to destination",
+                evt_tag_str("driver", self->super.super.id),
+                evt_tag_int("batch_size", self->batch_size));
+
+      _drop_batch(self);
+      _disconnect_and_suspend(self);
+      break;
+
+    case WORKER_INSERT_RESULT_ERROR:
+      self->retries.counter++;
+
+      if (self->retries.counter >= self->retries.max)
+        {
+          if (self->messages.retry_over && triggering_message)
+            self->messages.retry_over(self, triggering_message);
+
+          msg_error("Multiple failures while sending message(s) to destination, message(s) dropped",
+                    evt_tag_str("driver", self->super.super.id),
+                    evt_tag_int("number_of_retries", self->retries.max),
+                    evt_tag_int("batch_size", self->batch_size));
+
+          _drop_batch(self);
+        }
+      else
+        {
+          msg_error("Error occurred while trying to send a message, trying again",
+                    evt_tag_str("driver", self->super.super.id),
+                    evt_tag_int("retries", self->retries.counter),
+                    evt_tag_int("batch_size", self->batch_size));
+          _rewind_batch(self);
+          _disconnect_and_suspend(self);
+        }
+      break;
+
+    case WORKER_INSERT_RESULT_NOT_CONNECTED:
+      msg_debug("Server disconnected while preparing messages for sending, trying again",
+                evt_tag_str("driver", self->super.super.id),
+                evt_tag_int("batch_size", self->batch_size));
+      _rewind_batch(self);
+      _disconnect_and_suspend(self);
+      break;
+
+    case WORKER_INSERT_RESULT_EXPLICIT_ACK_MGMT:
+      /* we require the instance to use explicit calls to ack_messages/rewind_messages */
+      break;
+
+    case WORKER_INSERT_RESULT_SUCCESS:
+      stats_counter_inc(self->written_messages);
```

Comment:
When talking about statistics, should we increase any per-message counters with batch size? It might worth to cover this with some tests. e2e is great, but probably this could be even covered with a unit test.

Reply:
good catch. just pushed a couple of patches to at least fix the issues (both the statistics and also a related issue with $SEQNUM).

A dummy threaded destination would be very useful for testing, which is probably the thing I would do next, before actually making it possible to use multiple workers.

These are the fixes (I've kept them separate patches on top of my branch):
67785c68f776190c385aa8a992d56b8ce89769e8 logthrdestdrv: clean up sequence number increments
20242792dfdf67eba14c178aa4fc71528084ab26 logthrdestdrv: written_messages should get updated with the entire batch